include/picongpu/simulation/stage/Collision.hpp

```diff
@@ -47,19 +66,127 @@ namespace picongpu
+                    // A species temperature is assumed to be (2/3)<E_kin>
+                    if constexpr(particles::collision::calculateScreeningLength)
+                    {
+                        DataConnector& dc = Environment<>::get().DataConnector();
+                        using SpeciesSeq = picongpu::particles::collision::CollisionScreeningSpecies;
+                        constexpr uint32_t slot = picongpu::particles::collision::screeningLengthSlot;
+                        // get a FieldTmp for storring the result
+                        auto screeningLengthSquared = dc.get<FieldTmp>(FieldTmp::getUniqueId(slot), true);
+                        // The inverse of the Debye length squared is the sum over ScreeningInvSquared of each
+                        // species. Calculate it using the fold algorithm:
+                        particles::particleToGrid::FoldDeriveFields<
+                            CORE + BORDER,
+                            SpeciesSeq,
+                            pmacc::math::operation::Add,
+                            particles::particleToGrid::combinedAttributes::ScreeningInvSquared>{}(
+                            *screeningLengthSquared,
+                            currentStep,
+                            slot + 1u);
+
+                        // The Debye length is bound to be greater than the mean inter-atomic distance of the most
+                        // dense species ( this definition of the lower cut-off comes from [Perez 2012]).
+                        // To find this cut-off we need to find the highest density:
+                        auto maxDensity = dc.get<FieldTmp>(FieldTmp::getUniqueId(slot + 1u), true);
+                        particles::particleToGrid::FoldDeriveFields<
+                            CORE + BORDER,
+                            SpeciesSeq,
+                            pmacc::math::operation::Max,
+                            particles::particleToGrid::derivedAttributes::Density>{}(
+                            *maxDensity,
+                            currentStep,
+                            slot + 2u);
+
+                        // Invert the calculated value and apply the cut-off to get the squared debye length
+                        screeningLengthSquared->modifyByField<CORE + BORDER, GetSquaredScreeningLength>(*maxDensity);
+                        // the FieldTmp slot used for storing  the max density is not longer needed beyond this point
+                        maxDensity.reset();
+
+                        // write Debye length averaged over the whole simulation volume to a text file for debugging
+                        if constexpr(particles::collision::debugScreeningLength)
+                        {
+                            // calculate the average of square roots
+                            float_X sumRank = screeningLengthSquared->calculateSum<CORE + BORDER, true>();
```

Comment:
Please use `GolbalRecude` instead of the hand crafted version:

```C++
algorithms::GlobalReduce reduce(1024);
using D1Box = DataBoxDim1Access<typename GridBuffer<float3_X, simDim>::DataBoxType>;
D1Box d1access(buffer->getDeviceBuffer().getDataBox(), buffer->getGridLayout().getDataSpace());
int elements = buffer->getGridLayout().getDataSpace().productOfComponents();
auto result = reduce(pmacc::math::operation::Add(), d1access, elements);
```

If you need on device a handwritten functor you must split the command in a [local reduce](https://github.com/ComputationalRadiationPhysics/picongpu/blob/dev/include/pmacc/device/Reduce.hpp) and pass the result into an [MPI reduce](https://github.com/ComputationalRadiationPhysics/picongpu/blob/dev/include/pmacc/mpi/MPIReduce.hpp) with the operation `Add`